src/dev/typescript/project.ts

```diff
@@ -0,0 +1,69 @@
+import { readFileSync } from 'fs';
+import { basename, dirname, relative, resolve } from 'path';
+
+import { IMinimatch, Minimatch } from 'minimatch';
+import { parseConfigFileTextToJson } from 'typescript';
+
+import { File } from '../file';
+
+const ROOT_DIR = resolve(__dirname, '../../../');
+
+function makeMatchers(directory: string, patterns: string[]) {
+  return patterns.map(pattern => new Minimatch(resolve(directory, pattern), {
+    dot: true,
+  }));
+}
+
+export class Project {
+  public static fromConfig(path: string) {
+    const { error, config } = parseConfigFileTextToJson(path, readFileSync(path, 'utf8'));
+
+    if (error) {
+      throw error;
+    }
+
+    return new Project(path, config.files, config.include, config.exclude);
+  }
+
+  private include: IMinimatch[];
+  private exclude: IMinimatch[];
+
+  constructor(private tsConfigPath: string, files?: string[], include?: string[], exclude: string[] = []) {
+    if (files || !include) {
+      throw new Error(
+        'tsconfig.json files in the Kibana repo must use "include" keys and not "files"'
+      );
+    }
+
+    this.include = makeMatchers(dirname(tsConfigPath), include);
+    this.exclude = makeMatchers(dirname(tsConfigPath), exclude);
+  }
+
+  public getTsConfigPath() {
+    return this.tsConfigPath;
+  }
+
+  public isAbsolutePathSelected(path: string) {
+    return this.exclude.some(exc => exc.match(path))
+      ? false
+      : this.include.some(inc => inc.match(path));
+  }
+}
+
+export const TS_PROJECTS = [
+  Project.fromConfig(require.resolve('../../../tsconfig.json')),
+  Project.fromConfig(require.resolve('../../../packages/kbn-pm/tsconfig.json')),
+  Project.fromConfig(require.resolve('../../../packages/kbn-system-loader/tsconfig.json')),
```

Comment:
likewise if we're keeping it a static list, thoughts on adding a note that this list needs to be updated?

Reply:
I really didn't want to search the whole directory for `tsconfig.json` files over and over when the number of files is almost never going to change. Instead I put a little extra effort into https://github.com/spalger/kibana/blob/implement/typescript/src/dev/typescript/project.ts#L64 which will throw whenever someone tries to lint or commit a typescript file.

This way the tooling is snappier and only once in a blue moon will someone have to read an error message to understand how to fix the problem.

Reply:
If we limit the searching to the top level directories under `packages`, I suspect this is going to be essentially identical as far as humans can perceive. I think it's totally reasonable to hard code assumptions about where tsconfig files might live rather than allowing them on any arbitrary part of the file system, but specifying individual package directories doesn't seem necessary.

For one, I expect most of our packages to be typescript over time, so this list is going to converge on just being a copy of our portion of our file system. Plus, rather than packages always being something we link up through our dependency system, we're now baking in hardcoded implicit dependencies between specific packages and our core tooling. Having our typescript tooling be aware of the notion of how packages are stored in Kibana makes sense, but having it be aware that these specific packages have typescript just seems like unnecessary coupling.

What do you think?

Reply:
Yeah, a very limited glob set works for me.